### PR TITLE
Add initial version of article about technical debt

### PR DESCRIPTION
**Pull request by Ferroman**
Added article title image

### CODE REVIEW

File: _posts/2017-04-12-managing-technical-debt.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+modified: 2014-12-24
+tags: [code review, agile, process, qa]
+categories: [process]
+image:
+    feature: posts/tech_debt/title.jpg
+---
+
+# WHAT IS “TECHNICAL DEBT”?
+
+Martin Fowler [has a great explanation](https://martinfowler.com/bliki/TechnicalDebt.html) of this term and the problem behind it.  Many startups and new projects use quick-and-dirty solutions to achieve their goals as soon as possible, which in some point makes project complicated, slow, hard to change or implement a new features and full of bugs. I want to talk about most usual issues and how manage most effectively from my point of view.
```
Comment (intra1) [inline _posts/2017-04-12-managing-technical-debt.md]: ...Many startups and new projects use quick-and-dirty solutions to achieve their goals as soon as possible, which [at some point, in some cases, in some instances???] overcomplicates a project, slows it down, makes it hard to change or implement a new features and full of bugs.  I'd like to bring up the most common issues and how to manage them in the most effective way, from my point of view.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+Project with the big technical debt leading to make developers fear for changes. Any change can have unexpected effect and produce more bugs. And there is only one way to fearless development - it’s automated tests, which such projects usually don’t have. So technical debt is mostly about luck of control of the project quality.  There is other common issues in the project with big TD, but this is the most important one, and, what the worst thing here, very underestimated.
```
Comment (intra1) [inline _posts/2017-04-12-managing-technical-debt.md]: A project with big technical debt leading to the fear of change. [A project with big technical debt making developers fear change]. Any change can have an unexpected effect and produce more bugs. And there is only one way to a fearless development - it’s automated tests, which usually don’t exist on such projects. So technical debt is mostly about a lack of control of the project quality. There are other common issues on the projects with big TD, but this is the most important and, unfortunately, a heavily underestimated one.

**Review by intra1 — CHANGES_REQUESTED** [review]

File: _posts/2017-04-12-managing-technical-debt.md (hunk 1 above)
Comment (intra1) [inline _posts/2017-04-12-managing-technical-debt.md]: ...Many startups and new projects use quick-and-dirty solutions to achieve their goals as soon as possible, which [at some point, in some cases, in some instances???] overcomplicates a project, slows it down, makes it hard to change or implement new features and full of bugs. I'd like to bring up the most common issues and how to manage them in the most effective way, from my point of view.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ <!-- more -->
+ 
+# TL;DR: Plan
+1. **Create automatic backup for application users data.** We don’t want to corrupt user data.
+2. **Start using [commit strategy and workflow](http://nvie.com/posts/a-successful-git-branching-model/)** and make sure that everybody in a team understand it. There should be clear process to publish changes.
```
Comment (intra1) [inline _posts/2017-04-12-managing-technical-debt.md]: **Start using [commit strategy and workflow](http://nvie.com/posts/a-successful-git-branching-model/)** and make sure that everybody in the team understands it. There should be a clear process for publishing changes.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
+---
+layout: post
+title: Managing Technical Debt
+description: "Quite obvious plan to manage a technical debt."
```
Comment (intra1) [inline _posts/2017-04-12-managing-technical-debt.md]: The obvious plan how to manage technical debt

File: _posts/2017-04-12-managing-technical-debt.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+6. Configure Dev environment and onboarding manually
+7. Has no backup automation for data
+8. Has no data sampling for dev/QA purposes 
+9. Has a lot of “code smells” in source code, design by committee with a lot of hidden requirements due to lack of refactoring
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe each a little. 
+
+From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramastically with no chance to become better.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including [either without 'this' or ' this includes']: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramatically with no chance to become better.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+8. Has no data sampling for dev/QA purposes 
+9. Has a lot of “code smells” in source code, design by committee with a lot of hidden requirements due to lack of refactoring
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe each a little. 
+
+From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramastically with no chance to become better. 
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Without configuration management it’s very [likely to introduce][hard not to introduce] bugs on production

File: _posts/2017-04-12-managing-technical-debt.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe each a little. 
+
+From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramastically with no chance to become better. 
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users. 
+
+Without project requirements management impossible to create good deployment and onboarding process, so both can take a days or, sometimes, weeks.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Without project requirements management [it's] impossible to create good [what does it mean 'good'? efficient, quick, easy?] deployment and onboarding process, so both can take a days or, sometimes, weeks.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users. 
+
+Without project requirements management impossible to create good deployment and onboarding process, so both can take a days or, sometimes, weeks. 
+
+Without useful developer environment, developers will spend a lot of time trying to figure out is some particular issue are related to their local environment, or it’s a general issue; some thing can’t be run or tested on a local machine at all (for example delayed jobs, file storing in case when it’s configured to store on the third-party resources like S3, mail sending etc.); will be not able to reproduce bugs from other developers or production; it will be hard to implement configuration or new requirements without  breaking other team member processes;
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Without useful developer environment, developers will spend a lot of time trying to figure out [ if the particular issue is related to their local environment] or it’s a general issue; some thing[s] can’t be run or tested on a local machine at all (for example delayed jobs, file storing in case when it’s configured to store on the third-party resources like S3, mail sending etc.); will [not] be able to reproduce bugs from other developers or production; it will be hard to implement configuration or new requirements without breaking other team member processes;

File: _posts/2017-04-12-managing-technical-debt.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users. 
+
+Without project requirements management impossible to create good deployment and onboarding process, so both can take a days or, sometimes, weeks. 
+
+Without useful developer environment, developers will spend a lot of time trying to figure out is some particular issue are related to their local environment, or it’s a general issue; some thing can’t be run or tested on a local machine at all (for example delayed jobs, file storing in case when it’s configured to store on the third-party resources like S3, mail sending etc.); will be not able to reproduce bugs from other developers or production; it will be hard to implement configuration or new requirements without  breaking other team member processes; 
+
+Without simple and understandable deployment process team will have a hard time on delivering the fixes and new features, and this impact a application users which is critical. Technical dept in this process usually very time-consuming risky and stressful.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Without simple and understandable deployment process team will have a hard time [] delivering the fixes and new features, and this [can?] impact a[n] application users which is critical. Technical dept in this process usually very time-consuming risky and stressful.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+So, fixing the infrastructure issues is the key to manage the technical debt.  
+
+Data backing - it a most important part to decrease the risk. There is no way to make a lot of changes without mistakes, so team should not fear that their mistakes will result in data loss or corruption, which is, probably, the worst thing that can happen. 
+Managing the TD in code is more straightforward. Some solutions already covered by the best practices of used languages, tools and frameworks. To identify the problem and engineer the solution are not very difficult, and with the proper testing risk can be minimal. Refactoring, test creation and code reviews should be permanent part of the development process.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Managing the TD in [a] code is more straightforward. Some solutions already covered by the best practices [for?] of used languages, tools and frameworks.  [It's not very difficult to identify the problem and engineer the solution], and with the proper testing risk can be minimal. Refactoring, test creation and code reviews should be permanent part of the development process.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+Without simple and understandable deployment process team will have a hard time on delivering the fixes and new features, and this impact a application users which is critical. Technical dept in this process usually very time-consuming risky and stressful.  
+
+Without testing infrastructure team can’t create the tests. It’s not necessary to became crazy and move all resources to create tests for existing code, but developers have to has possibility to creating the tests when they can and run in a regular bases. Without tests developers can’t effectively refactor code or implement a new features without introducing new bugs, so they should do all manual testing which is very ineffective, time-consuming and bad for team morale. Having this infrastructure will give the opportunity to became project health better over the time. Without it this will likely never happen.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Without testing infrastructure team can’t create the tests. It’s not necessary to bec[o]me crazy and move all resources to create tests for existing code, but developers have to [have a] possibility to create a tests when they can and run in a regular bases. Without tests developers can’t effectively refactor code or implement a new features without introducing new bugs, so they should do all manual testing which is very ineffective, time-consuming and bad for team morale. Having this infrastructure will give the opportunity to [make] project's health better over the time. Without it this will likely never happen.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+1. Make user data safe
+2. Isolate and unify environment and configurations to avoid confusions
+3. Add tests infrastructure
+4. Make development more safe by adding high-level tests
+5. Make delivery process as simple as possible
+6. Cover changed code with tests
+7. Make real debt visible to everyone
+8. Use simple rules to make things clear  
+
+Hope it helps!
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md:103]: I think it's not ready yet, it sounds like prototype, a starting point for an article. For me it was hard to follow your key points. 
Stepping aside: after my small experience with CI I would mention some code complexity measuring tools, code quality tools. As they can decrease TD related to codebase. for e.g (python: Radon, pep8, pylint)

**Review by akozl — CHANGES_REQUESTED** [review]

File: _posts/2017-04-12-managing-technical-debt.md (hunk 7 above)
Comment (Ferroman) [reply _posts/2017-04-12-managing-technical-debt.md]: > what does it mean 'good'? efficient, quick, easy?

all of them

File: _posts/2017-04-12-managing-technical-debt.md (hunk 12 above)
Comment (Ferroman) [reply _posts/2017-04-12-managing-technical-debt.md:103]: It's, probably, structure. I'll think how to restructure. 
Yeah I thought about whese measurment, but they are not very helpful on project with the big TD, these tools are mostly about to prevention TD. They are useful when big isues are already fixed, but you're right, I should mention it.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+Without simple and understandable deployment process team will have a hard time delivering the fixes and new features, 
+and this can impact an application users which is critical. Technical dept in this process usually very time-consuming risky and stressful. 
+
+Code issues and data sampling issues is simply consequences of infrastructure and process issues. Test, code reviews and code quality measurement tools will help to deal with that.  
+
+One more thing, which, I believe,  is very important to successful reduce of TD. Technical debt does not grow overnight. Sometimes it takes a years before people start thinking about to do something with, and it’s usually because they didn’t saw the whole picture to the very end, and start looking at it when any change became taking weeks and bugs and complaints numbers growth exponentially. This often because of issues organization for tech team.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: One more thing, which, I believe,  is very important to successful reduce of TD. Technical debt does not grow overnight. Sometimes it takes **[]** years before people start thinking about **[doing something with it],** usually because they didn’t saw the whole picture **[till]** the very end, and start looking at it when any change became taking weeks and bugs and complaints numbers growth exponentially. [**This often because of issues organization for tech team.** _I don't get it. Maybe rephrase?_]

File: _posts/2017-04-12-managing-technical-debt.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+Code issues and data sampling issues is simply consequences of infrastructure and process issues. Test, code reviews and code quality measurement tools will help to deal with that.  
+
+One more thing, which, I believe,  is very important to successful reduce of TD. Technical debt does not grow overnight. Sometimes it takes a years before people start thinking about to do something with, and it’s usually because they didn’t saw the whole picture to the very end, and start looking at it when any change became taking weeks and bugs and complaints numbers growth exponentially. This often because of issues organization for tech team. 
+Leads organize bugs/stories/requests/tasks independently into different lists/projects, which hide the real scope of work. To avoid this, team should collect all technical tasks, bugs and issues in the one list. It will be a big list, but it will show how are you really deal with existing Technical Debt and implement features stories in the same time. Keeping things in different lists/boards/projects only hide the real situation.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: Leads organize bugs/stories/requests/tasks independently into different lists/projects, which hide the real scope of work. To avoid this, team should collect all technical tasks, bugs and issues in the one list. It will be a big list, but it will show how **[]** you really deal with existing Technical Debt and implement features stories in the same time. Keeping things in different lists/boards/projects only hide the real situation.

File: _posts/2017-04-12-managing-technical-debt.md [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+6. Do not track requirements
+7. Configure Dev environment and onboarding manually
+8. Has no data sampling for dev/QA purposes 
+9. Has a lot of “code smells” in source code, design by committee with a lot of hidden requirements due to lack of refactoring
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe them a little. 
+
+Data backing - it a most important part to decrease the risk. There is no way to make a lot of changes without mistakes, so team should not fear that their mistakes will result in data loss or corruption, which is, probably, the worst thing that can happen.
```
Comment (akozl) [inline _posts/2017-04-12-managing-technical-debt.md]: A paragraph missing?
